File: legacy/python/spc_up/spca/load_tables.py
"""Load SPCA domain tables from YAML files."""

from __future__ import annotations

from functools import lru_cache
from pathlib import Path
from typing import Any

import yaml

TABELAS_DIR = Path(__file__).parent / "tabelas"


def _normalize_code(code: str | int) -> str:
    return str(code).strip()


def _lookup_entry(table: dict[Any, Any], code: str | int) -> dict[str, Any] | None:
    key = _normalize_code(code)
    entry = table.get(key)
    if entry is not None:
        return entry
    if key.isdigit():
        return table.get(int(key))
    return None


@lru_cache
def _load_table(filename: str) -> dict[Any, Any]:
    path = TABELAS_DIR / filename
    with path.open(encoding="utf-8") as handle:
        data = yaml.safe_load(handle)
    if not isinstance(data, dict):
        raise ValueError(f"Expected mapping in {filename}")
    return data


def _format_label(code: str | int, filename: str) -> str:
    key = _normalize_code(code)
    entry = _lookup_entry(_load_table(filename), code)
    if entry is None:
        return key
    descricao = entry.get("descricao", "")
    if descricao:
        return f"{key} - {descricao}"
    return key


def get_classificacao_label(code: str | int) -> str:
    """Return human-readable label for a classificacao_receita code."""
    return _format_label(code, "classificacao_receita.yaml")


def get_gasto_label(code: str | int) -> str:
    """Return human-readable label for a cdDescricaoGasto code."""
    return _format_label(code, "codigos_gasto.yaml")

File: packages/spca/src/load-tables.ts
import { readFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import yaml from "js-yaml";

const TABELAS_DIR = join(dirname(fileURLToPath(import.meta.url)), "..", "tabelas");

type TableEntry = { descricao?: string };
type Table = Record<string | number, TableEntry>;

const tableCache = new Map<string, Table>();

function normalizeCode(code: string | number): string {
  return String(code).trim();
}

function loadTable(filename: string): Table {
  const cached = tableCache.get(filename);
  if (cached) {
    return cached;
  }

  const path = join(TABELAS_DIR, filename);
  const data = yaml.load(readFileSync(path, "utf-8"));

  if (typeof data !== "object" || data === null || Array.isArray(data)) {
    throw new Error(`Expected mapping in ${filename}`);
  }

  const table = data as Table;
  tableCache.set(filename, table);
  return table;
}

function lookupEntry(table: Table, code: string | number): TableEntry | null {
  const key = normalizeCode(code);
  const entry = table[key];
  if (entry !== undefined) {
    return entry;
  }
  if (/^\d+$/.test(key)) {
    const numeric = table[Number(key)];
    if (numeric !== undefined) {
      return numeric;
    }
  }
  return null;
}

function formatLabel(code: string | number, filename: string): string {
  const key = normalizeCode(code);
  const entry = lookupEntry(loadTable(filename), code);
  if (entry === null) {
    return key;
  }
  const descricao = entry.descricao ?? "";
  if (descricao) {
    return `${key} - ${descricao}`;
  }
  return key;
}

/** Return human-readable label for a classificacao_receita code. */
export function getClassificacaoLabel(code: string | number): string {
  return formatLabel(code, "classificacao_receita.yaml");
}

/** Return human-readable label for a cdDescricaoGasto code. */
export function getGastoLabel(code: string | number): string {
  return formatLabel(code, "codigos_gasto.yaml");
}
